src/HX711Multi.py

```diff
@@ -0,0 +1,238 @@
+
+read()
+    Gets a reading from all HX711s. NOTE: isReady() must return True before
+    calling read() or it will fail.
+
+    Returns a list of integer readings with offsets applied from the HX711s
+    if isReady(), otherwise returns False.
+
+read_raw()
+    Gets raw integer readings from all HX711s. NOTE: isReady() must return True
+    before calling read_raw() or it will fail.
+
+    Returns a list of integer readings from the HX711s if isReady(),
+    otherwise returns False.
+
+"""
+
+class HX711_Multi:
+
+    __PD_SCK_pin = None
+    __DOUT_pins = []
+    __gain = None
+    __debug_enabled = None
+    __offsets = [] 
+    __timeout = 300
+    __num_of_HX711s = None
+
+    def __init__(self, 
+                data_pins: List[int],
+                clock_pin: int,
+                gain: int = 128,
+                debug: bool = False):
+
+        self.__PD_SCK_pin = clock_pin
+        self.__DOUT_pins = data_pins
+        self.__debug_enabled = debug
+        self.__num_of_HX711s = len(self.__DOUT_pins)
+
+        GPIO.setwarnings(False)
+        GPIO.setmode(GPIO.BCM)
+
+        # Set the pins as outputs
+        GPIO.setup(self.__PD_SCK_pin, GPIO.OUT)
+        for pin in self.__DOUT_pins:
+            GPIO.setup(pin, GPIO.IN)
+            
+        self.__setGain(gain)
+        self.__powerUp()
+
+    def isReady(self):
+
+        # DOUT pin goes low (0V) when it is ready to send a reading
+        for data_pin in self.__DOUT_pins:
+            if GPIO.input(data_pin) == GPIO.HIGH:
+                if self.__debug_enabled:
+                    print("HX711 on pin", data_pin, "not ready")
+                    continue
+                return False
+                
+        return True
+
+    def tare(self, num_of_samples: int, max_std_dev_from_mean: float):
+
+        if num_of_samples < 100:
+            print("Cannot tare with less than 100 samples!")
+            return False
+        
+        samples = []
+        while len(samples) < num_of_samples:
+            if self.isReady():
+                samples.append(self.readRaw())
+
+        """
+            Do some fat stats to reject outlier data points based on
+            https://www.adamsmith.haus/python/answers/how-to-remove-
+            outliers-from-a-numpy-array-in-python
+        """
+
+        samples = np.array(samples).T
+        means = np.mean(samples, axis=1).reshape(-1,1)
+        std_devs = np.std(samples, axis=1).reshape(-1,1)
+        dists_from_means = abs(samples - means)
+
+        # True/false mask to extract non outlier values
+        no_outliers = dists_from_means <= (max_std_dev_from_mean * std_devs)
+        clean_data = [row[no_outliers[i]] for i,row in enumerate(samples)]
+
+        clean_data_length = 0
+        for data_list in clean_data:
+            clean_data_length += len(data_list)
+
+        if len(clean_data) < 0.8*len(samples):
+            if self.__debug_enabled:
+                print("Failed to tare. Excessive deviations measured.")
```

Comment:
same thing as before with the assertions. Do that for all debugging print statments

Reply:
See other reply